internal/health/condition/check_data_volumes_ready.go

```diff
@@ -0,0 +1,65 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package condition
+
+import (
+	"context"
+	"fmt"
+
+	druidv1alpha1 "github.com/gardener/etcd-druid/api/v1alpha1"
+	"github.com/gardener/etcd-druid/internal/utils"
+
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type dataVolumesReady struct {
+	cl client.Client
+}
+
+func (d *dataVolumesReady) Check(ctx context.Context, etcd druidv1alpha1.Etcd) Result {
```

Comment:
pod status and statefulset status are already capturing the status of volumes, then why to capture this again in etcd resource status ... to me it seems like a redundant with little value added.

Reply:
we are here just reading the statefulset status and copying it to the etcd custom resource status. How it's helping as right now etcd controller is not taking action by capturing this data in its own status

Reply:
> pod status and statefulset status are already capturing the status of volumes

No, they don't. All information about volume attachment and detachment are captured as events in the cluster, and it's painful for a human operator to check all events just to figure out why the etcd isn't getting ready. A condition `DataVolumesReady` provides this as a very simple and easy-to-read condition on the Etcd resource status itself. Previously, during etcd spec reconciliation, etcd controller would wait for some time, capture these events, check if there's a volume attachment error, and then return that as a reconciliation error, which was incorrect as well as time-consuming, and resulted in long reconciliation loops. With the asynchronous condition `DataVolumesReady`, we simply eliminated the need for etcd reconciliation reconciler to wait for the sts to be deployed, and made reconciliations extremely quick.

> we are here just reading the statefulset status and copying it to the etcd custom resource status

No. As mentioned above, we fetch the events corresponding to the sts, then check which one of those are warning events, and then capture the warning (error) messages as the condition message for this condition.

> right now etcd controller is not taking action by capturing this data in its own status

Right now, this is providing valuable diagnostic information to any human operator, to understand why an etcd might be failing to get ready. Additionally, in the future, druid could even take action is the sts encounters a volume attach error, like possibly restarting the pod to force it to be scheduled to a different node, or even take actions on the volumeattachment objects. These are farfetched ideas, but would be extremely helpful for environments/providers who frequently face volume detach/attach errors. Since etcd is a stateful workload, and depends heavily on the volumes, this condition becomes very important.

Reply:
ok, seems fine